### ISSUE TITLE
Parallel version of miniapps/get-values fails to get all values 

### ISSUE THREAD

**Issue by jchan192** [mfem/mfem#2785]
Dear MFEM community,

As a quick test, I run ex16p with 
`mpirun -np 4 ex16p -m ../data/amr-quad.mesh -o 4 -rs 0 -rp 0 --visit-datafiles`
which generates outputfiles Example16-Parallel_0000xx where xx = 00, 05, 10, .. , 50
I have checked the snapshots with parallel version of glvis. The resulting figure looks fine.

To get values at 4 positions (0.25 0.25), (0.25 0.75), (0.75 0.25), (0.75 0.75),
I try to use the miniapps/tools/get-values 
`mpirun -np 4 ./../miniapps/tools/get-values -r Example16-Parallel -c 0 -p "0.25 0.25 0.25 0.75 0.75 0.25 0.75 0.75" -fn All`

But it only gets the value at [0.25,0.25], and failed to get values at other positions.
here is the following error 

```
MFEM Warning: 3 points were not found
 ... in function: virtual int mfem::Mesh::FindPoints(mfem::DenseMatrix&, mfem::Array<int>&, mfem::Array<mfem::IntegrationPoint>&, bool, mfem::InverseElementTransformation*)
 ... in file: mesh/mesh.cpp:11704

Found 1 points.
# Number of fields
1
# Legend
# "Index" "Location":2
2
# Number of points
1
0 0.25 0.25
```
I believe the problem is that get-values.cpp fails to load the other 3 sub-domains / meshes. 
Thank you for your time, and I really appreciate it if someone know a quick fix to this issue.

**Comment by kmittal2** [comment]
Hi @jchan192,

Seems like you are right. Can you try to use `GSLIB::FindPoints` for interpolating the solution at the desired set of points?
See the install instructions [here](https://github.com/mfem/mfem/blob/master/INSTALL#L719). Once you have built MFEM in GSLIB, have a detailed look at the GSLIB interpolation example, [pfindpts](https://github.com/mfem/mfem/blob/master/miniapps/gslib/pfindpts.cpp).
Adding something like this at the bottom of ex16p once you have MFEM built with GSLIB should do the trick:
```
   // Set mesh curvature (required by GSLIB)
   pmesh->SetCurvature(1);
   // Define number of points to be found
   int pts_cnt = 2;
   // Define vector that will hold the number of points (ordering is x1,x2,x3...xN,y1,y2,y3...yN)
   Vector vxyz(pts_cnt * pmesh->Dimension());
   vxyz(0) = 0.25; //x1
   vxyz(0+pts_cnt) = 0.25; //y1
   vxyz(1) = 0.75; //x2
   vxyz(1+pts_cnt) = 0.75; //y2
   if (myid != 0) {
       vxyz.Destroy();
       pts_cnt = 0; //We let Node 0 find all the points. Alternatively one could split the points to be found across all ranks.
   }
   // Define the vector for interpolated values
   Vector interp_vals(pts_cnt);
   // Define FindPointsGSLIB object for interpolation
   FindPointsGSLIB finder(MPI_COMM_WORLD);
   finder.Setup(*pmesh);
   finder.Interpolate(vxyz, u_gf, interp_vals);
   // Free data allocation done inside FindPoints
   finder.FreeData();
```

Hope this helps,
Ketan

**Comment by jchan192** [comment]
Thanks Ketan for the reply.
Maybe I will try to do it later, because it will take me some time to install GSLIB.
I really appreciate your help.

**Comment by jchan192** [comment]
As an alternative solution, I try to use `gf.GetNodalValues(int i)` and `pmesh.GetVertex(int i)` or `pmesh.GetVertices(vector)`  to get the solution and coordinates on the mesh, but the problem is both methods give repeated coordinates, but I expect vertices to be unique.

My parallelized pmesh is a periodic 3d Box, generated by the following

```
   double BoxL = 2.0;
   int    BoxN = 8;
   Mesh mmesh = Mesh::MakeCartesian3D(BoxN,BoxN,BoxN, Element::HEXAHEDRON, BoxL,BoxL,BoxL);

   std::vector<Vector> translations =
     {
         Vector({BoxL, 0.0, 0.0}),
         Vector({0.0, BoxL, 0.0}),
         Vector({0.0, 0.0, BoxL})
     };
   mmesh = Mesh::MakePeriodic(mmesh,
                              mmesh.CreatePeriodicVertexMapping(translations));
   Mesh mesh = Mesh(mmesh);
   mesh.EnsureNCMesh();
   mesh.Finalize(true);
   ParMesh pmesh = ParMesh(MPI_COMM_WORLD, mesh);

```
I print out the coordinates in the following way.

```
       int NE = pmesh.GetNE();
       Vector savecoord(3*NE);
       Vector savegf(NE);
       gf.GetNodalValues(savegf,1);
       pmesh.GetVertices(savecoord);

       for (int i = 0; i<NE;i++){
         std::cout << savecoord[3*i]  << " "
                   << savecoord[3*i+1]  << " "
                   << savecoord[3*i+2]  << " ";
                   << savegf[i]  << "\n";

```

**Comment by v-dobrev** [comment]
@kmittal2, were you able to reproduce the issue with the `get-values` miniapp? If there is indeed an issue with that, we should try to resolve it.

@mlstowell, maybe you can take a look?

**Comment by kmittal2** [comment]
Yes, I was able to reproduce the issue with `get-values`. 

The issue is that the `Load` method in `DataCollection` as indicated by the Class description, does not work for parallel meshes. So the mesh gets loaded as a serial mesh on each rank, and with FindPoints, each sub-domains finds one of the 4 points that they have. 
Once `FindPoints` is done, we can use the `elem_ids` returned by the method to determine which points were found on each rank by checking for `elem_ids[k] > -1`. cc: @jchan192

**Comment by kmittal2** [comment]
Yes, I was able to reproduce the issue with `get-values`. 

The issue is that the `Load` method in `DataCollection` as indicated by the Class description, does not work for parallel meshes. So the mesh gets loaded as a serial mesh on each rank, and with FindPoints, each sub-domains finds one of the 4 points that they have. 
Once `FindPoints` is done, we can use the `elem_ids` returned by the method to determine which points were found on each rank by checking for `elem_ids[k] > -1`. cc: @jchan192

**Comment by v-dobrev** [comment]
Actually, it seems like `VisItDataCollection` should support saving and reading parallel meshes in parallel format. I suspect the problem is that by default we save parallel meshes in the serial format from each rank. However, switching to saving in the parallel format should be easy, just add
```c++
      visit_dc.SetFormat(DataCollection::PARALLEL_FORMAT);
```
to the example before the call to `visit_dc.Save()`.

**Comment by jchan192** [comment]
adding    `visit_dc.SetFormat(DataCollection::PARALLEL_FORMAT);` before `visit_dc.Save()`
erased the warning `MFEM Warning: 3 points were not found` of `get-values.cpp`, but still only gives the value at [0.25,0.25].

I think the code `get-values.cpp` itself still needs some modification.

**Comment by jchan192** [comment]
I managed to print out the values when I changed the `mfem:out` to `std::cout`. Therefore, `get-value.cpp` does correctly load the parallel meshes after the above changes. I have modified the code to save the values into binary file, because saving binary data is easier in MPI than text files.  So I would not say the problem is completely solved, but Im happy for now.  Thanks for the help!

-------------------------------------------------------------------------------------------------------------------------- 
Please let me ask another related question here.
I try to get values and also the coordinates at the vertices of the mesh, but since my mesh is periodical, the resulting coordinates from `pmesh.GetVertices(savecoord)` contain repeated values (lots of them are duplicate). This is not convenient for post-processing the final solution. I wonder if anyone knows how to get non-duplicate coordinates of the periodical mesh? Thanks again.

**Comment by kmittal2** [comment]
@jchan192,

You are seeing duplicated values because they are not being printed in the right order. You should print them out something like this:

```
  const int nverts = savecoord.Size()/pmesh.Dimension();
  for (int i = 0; i<nverts; i++){
    std::cout << i << " " << savecoord[i]  << " "
              << savecoord[i+nverts]  << " "
              << savecoord[i+2*nverts]  << " \n";
  }
```

Note: the number of vertices is not the same as the number of elements as indicated in the code snippet you pasted above.

Ketan

**Comment by jchan192** [comment]
Thanks @kmittal2  for the reply.

I used your code, but it still produces duplicated coordinates. Here is an output showing the first printed 12 points with  when run with only 1 node `mpirun -n p1 `  I still think the problem comes from the periodical property of the mesh, because when I commented the `translation` part of the code, each coordinates then become unique. However, my problem requires periodical  boundary, so I still want to solve the issue. 
```
0 1 1 1                                                                                                                                                                                                            
1 0.25 1 1                                                                                                                                                                                                         
2 0.5 1 1                                                                                                                                                                                                          
3 0.75 1 1                                                                                                                                                                                                         
4 1 1 1                                                                                                                                                                                                            
5 1.25 1 1                                                                                                                                                                                                         
6 1.5 1 1                                                                                                                                                                                                          
7 1.75 1 1                                                                                                                                                                                                         
8 1 0.25 1                                                                                                                                                                                                         
9 0.25 0.25 1                                                                                                                                                                                                      
10 0.5 0.25 1                                                                                                                                                                                                      
11 0.75 0.25 1                                                                                                                                                                                                     
12 1 0.25 1   
```

**Comment by kmittal2** [comment]
Hi @jchan192,

Yes, I was able to verify that this is indeed a problem for periodic meshes.
Essentially, what happens is that when we make a mesh periodic, internally we set its [curvature to 1 ](https://github.com/mfem/mfem/blob/master/mesh/mesh.cpp#L4621) for a linear mesh (thus `Nodes != Null` anymore). When we use this periodic mesh to define a `ParMesh`, [`SetVerticesFromNodes()`](https://github.com/mfem/mfem/blob/master/mesh/pmesh.cpp#L290) is called which averages the coordinates of periodic vertices. As a result, the vertices being printed out for periodic vertices are innacurate.

I am not sure yet of the right way to fix this issue.

Ketan
